// @sap/async-xsjs/lib/xsjs/trace/trace.js
'use strict';

var logging = require('../../logging');


exports.createTrace = function (options) {
  var id = options.id;
  var location = options.location;
  var req = options.req;

  var tracer;
  if (id) {
    tracer = logging.appContext.createLogContext({ id: id }).getTracer(location);
  } else if (req && req.loggingContext) {
    tracer = req.loggingContext.getTracer(location);
  } else if (location) {
    tracer = logging.appContext.createLogContext({ id: '' }).getTracer(location);
  } else {
    tracer = logging.tracer;
  }
  return buildTraceObject(tracer);
};

function buildTraceObject(tracer) {
  return {
    debug: tracer.debug.bind(tracer),
    info: tracer.info.bind(tracer),
    warning: tracer.warning.bind(tracer),
    warn: tracer.warning.bind(tracer),
    error: tracer.error.bind(tracer),
    fatal: tracer.fatal.bind(tracer),

    isDebugEnabled: tracer.isEnabled.bind(tracer, 'debug'),
    isInfoEnabled: tracer.isEnabled.bind(tracer, 'info'),
    isWarningEnabled: tracer.isEnabled.bind(tracer, 'warning'),
    isErrorEnabled: tracer.isEnabled.bind(tracer, 'error'),
    isFatalEnabled: tracer.isEnabled.bind(tracer, 'fatal'),

    _tracer: tracer
  };
}
