// @sap/xsodata/lib/sql/dataCollectorPost.js
'use strict';

//Include
const sqlStatement = require('./sqlStatement');
const statementProcessor = require('./statementProcessor');
const utils = require('../utils/utils');
const SqlError = require('../utils/errors/sqlError');

exports.moveRecordNV_ToSelectStm = function (context, asyncDone) {
    context.logger.silly('dataCollectorPost', 'moveRecordNV_ToSelectStm');
    //get dbSegment for inserting the payload
    var dbSeg = context.oData.dbSegmentLast;

    var stm = dbSeg.sql.stmContainer.insertTmp;
    if (!stm) {
        return asyncDone("Internal server error", context);
    }

    var max = dbSeg._recordNV.length;
    var i;
    for (i = 0; i < max; i++) {
        var nv = dbSeg._recordNV[i];
        stm.setValue(nv.name, nv.value, dbSeg.entityType.propertiesMap[nv.name].DATA_TYPE_NAME); // OK
    }
    return asyncDone(null, context);
};


exports.createTmpTables = function (context, asyncDone) {
    context.logger.silly('dataCollectorPost', 'createTmpTables');
    var stm = context.oData.dbSegmentLast.sql.stmContainer.createTmp;
    return statementProcessor.execStmsDirectlyNoResult(context, stm, asyncDone);
};

exports.checkAutoValues = null;

exports.checkForAutoKeyGenUsage = function (context, asyncDone) {
    context.logger.silly('dataCollectorPost', 'checkForAutoKeyGenUsage');

    var dbSeg = context.oData.dbSegmentLast;
    if (!exports.checkAutoValues && !dbSeg.entityType.hasAdmindata()) {
        return asyncDone(null, context);
    }


    var stm = dbSeg.sql.stmContainer.insertTmp;

    var table = dbSeg._DB_TableName;

    var appData = {};
    var i;
    var nv = stm.nv;
    var name = '';

    const createBy = dbSeg.entityType.getAddAdmindata('create', 'by');
    const createAt = dbSeg.entityType.getAddAdmindata('create', 'at');
    const modifyBy = dbSeg.entityType.getAddAdmindata('modify', 'by');
    const modifyAt = dbSeg.entityType.getAddAdmindata('modify', 'at');


    for (i = 0; i < nv.length; i++) {

        name = nv[i].name.toSqlHana(null, null, { withoutTable: true });
        name = name.substr(1, name.length - 2);

        if (createBy && name === 'CREATED_BY') {
            nv[i].overwriteValue = new sqlStatement.Value(context.userName);
        }
        if (createAt && name === 'CREATED_AT') {
            nv[i].formula = 'NOW()';
        }
        if (modifyBy && name === 'MODIFIED_BY') {
            nv[i].overwriteValue = new sqlStatement.Value(context.userName);
        }
        if (modifyAt && name === 'MODIFIED_AT') {
            nv[i].formula = 'NOW()';
        }


        // build application api structure

        appData[name] = {
            value: nv[i].value,
            sequence: null
        };
    }

    if (!exports.checkAutoValues) {
        return asyncDone(null, context);
    }

    return exports.checkAutoValues(table, appData, function (err) {
        var name;
        for (i = 0; i < nv.length; i++) {
            name = nv[i].name.toSqlHana(null, null, { withoutTable: true });
            name = name.substr(1, name.length - 2);
            if (appData[name].sequence) {
                nv[i].sequence = appData[name].sequence;
            }
        }
        return asyncDone(err, context);
    });
};

exports.insertPayloadIntoTempTable = function (context, asyncDone) {
    context.logger.silly('dataCollectorPost', 'insertPayloadIntoTempTable');
    var dbSeg = context.oData.dbSegmentLast;

    statementProcessor.execStmsAsPreparedNoResult(context, dbSeg.sql.stmContainer.insertTmp, asyncDone);
};

exports.insertTmpTableToRealTable = function (context, asyncDone) {
    context.logger.silly('dataCollectorPost', 'insertTmpTableToRealTable');
    var dbSeg = context.oData.dbSegmentLast;
    statementProcessor.execStmsAsPreparedNoResult(context, dbSeg.sql.stmContainer.insertReal, asyncDone);
};

exports.selectData = function (context, asyncDone) {
    context.logger.silly('dataCollectorPost', 'selectData');
    var dbSeg = context.oData.dbSegmentLast;
    var stm = dbSeg.sql.stmContainer.select;

    return statementProcessor.execStmsAsPrepared(context, stm, (err,context, resultSet) => {
        const rows = resultSet[0];
        if (err) {
            return asyncDone(new SqlError(context, err), context);
        }
        dbSeg.sql.rows = rows;
        if (rows.length === 1 && utils.isETagHeaderRequired(context, dbSeg)) {
            dbSeg.etagHeader = rows[0].__etag;
        }
        context.logger.silly('selectData', 'rows: ' + JSON.stringify(rows.length));
        return asyncDone(null, context);
    });
};


exports.commit = function (context, asyncDone) {
    context.logger.debug('dataCollectorPost', 'commit');
    var client = context.db.client;
    client.commit(function (err) {
        if (err) {
            context.logger.info('SQL Exec', 'Commit Error: \n' + JSON.stringify(err));
            return asyncDone(err, context);
        }
        return asyncDone(null, context);
    });
};

/**
 * Executes truncation of temporary created tables
 *
 * @param {Object} context The xsodata context
 * @param {Function} asyncDone async waterfall callback
 */
exports.truncateTempTables = function (context, asyncDone) {
    context.logger.silly('dataCollectorPost', 'truncateTempTables');
    return statementProcessor.execTempTableStatements(context, context.sql.container.createTmpTruncate, asyncDone);
};

/**
 * Executes deletion of temporary created tables
 *
 * @param {Object} context The xsodata context
 * @param {Function} asyncDone async waterfall callback
 */
exports.dropTempTables = function (context, asyncDone) {
    context.logger.silly('dataCollectorPost', 'dropTempTables');
    statementProcessor.execTempTableStatements(context, context.sql.container.createTmpDrop, asyncDone);
};
