// @sap/async-xsjs/lib/xsjs/constants.js
'use strict';

exports.httpStatusCodes = {
  CONTINUE: 100,
  SWITCH_PROTOCOL: 101,

  OK: 200,
  CREATED: 201,
  ACCEPTED: 202,
  NON_AUTHORITATIVE: 203,
  NO_CONTENT: 204,
  RESET_CONTENT: 205,
  PARTIAL_CONTENT: 206,

  MULTIPLE_CHOICES: 300,
  MOVED_PERMANENTLY: 301,
  FOUND: 302,
  SEE_OTHER: 303,
  NOT_MODIFIED: 304,
  USE_PROXY: 305,
  TEMPORARY_REDIRECT: 307,

  BAD_REQUEST: 400,
  UNAUTHORIZED: 401,
  PAYMENT_REQUIRED: 402,
  FORBIDDEN: 403,
  NOT_FOUND: 404,
  METHOD_NOT_ALLOWED: 405,
  NOT_ACCEPTABLE: 406,
  PROXY_AUTH_REQUIRED: 407,
  REQUEST_TIMEOUT: 408,
  CONFLICT: 409,
  GONE: 410,
  LENGTH_REQUIRED: 411,
  PRECONDITION_FAILED: 412,
  REQUEST_ENTITY_TOO_LARGE: 413,
  REQUEST_URI_TOO_LONG: 414,
  UNSUPPORTED_MEDIA_TYPE: 415,
  REQUESTED_RANGE_NOT_SATISFIABLE: 416,
  EXPECTATION_FAILED: 417,

  INTERNAL_SERVER_ERROR: 500,
  NOT_YET_IMPLEMENTED: 501,
  BAD_GATEWAY: 502,
  SERVICE_UNAVAILABLE: 503,
  GATEWAY_TIMEOUT: 504,
  HTTP_VERSION_NOT_SUPPORTED: 505
};

exports.webTypes = {
  INVALID: -1,
  OPTIONS: 0,
  GET: 1,
  HEAD: 2,
  POST: 3,
  PUT: 4,
  DEL: 5,
  TRACE: 6,
  CONNECT: 7,
  PATCH: 8
};

exports.WEB = {
  MESSAGES : {
    LINE_BREAK : '\r\n'
  },
  MESSAGE_TYPE : {
    REQUEST : 'request',
    RESPONSE : 'response'
  }
};
