// @sap/async-xsjs/lib/xsjs/jobs/Schedule.js
'use strict';

var assert = require('assert');

var Logs = require('./Logs');

module.exports = Schedule;

function Schedule(scheduler, description, parameter, xscron, active, jobId, scheduleId) {
  assert(scheduler, 'Missing scheduler');
  assert(scheduleId, ' Missing scheduleId');
  assert(jobId, 'Missing jobId');
  var self = this;

  Object.defineProperty(this, 'description', {
    enumerable: true,
    get: function () {
      return description;
    },
    set: async function (value) {
      if (typeof value !== 'string') {
        throw new Error('Incorrect type for "description"');
      }
      await updateJobSchedulePromise(scheduler, {
        jobId: jobId,
        scheduleId: scheduleId,
        schedule: {
          description: value
        }
      });
      description = value;
    }
  });
  Object.defineProperty(this, 'active', {
    enumerable: true,
    get: function () {
      return active;
    },
    set: async (value) => {
      if (typeof value !== 'boolean') {
        throw new Error('Incorrect type for "Active"');
      }
      await updateJobSchedulePromise(scheduler, {
        jobId: jobId,
        scheduleId: scheduleId,
        schedule: {
          active: value
        }
      });
      active = value;
    }
  });

  Object.defineProperty(this, 'parameter', {
    enumerable: true,
    get: function () {
      return parameter;
    },
    set: async (value) => {
      if (typeof value !== 'string' && typeof value !== 'object') {
        throw new Error('Incorrect type for "parameter"');
      }
      await updateJobSchedulePromise(scheduler, {
        jobId: jobId,
        scheduleId: scheduleId,
        schedule: {
          data: (typeof value === 'string') ? JSON.parse(value) : value
        }
      });
      parameter = value;
    }
  });
  Object.defineProperty(this, 'xscron', {
    enumerable: true,
    get: function () {
      return xscron;
    },
    set: async function (value) {
      if (typeof value !== 'string') {
        throw new Error('Incorrect type for "xscron"');
      }
      await updateJobSchedulePromise(scheduler, {
        jobId: jobId,
        scheduleId: scheduleId,
        schedule: {
          cron: value
        }
      });
      xscron = value;
    }
  });

  return new Promise((resolve, reject) => {
    var logsPromise = new Logs(scheduler, scheduleId, jobId);
    logsPromise.then((logs) => {
      self.logs = logs;
      resolve(self);
    }).catch((err) => {
      reject(err); });
  });
}

function updateJobSchedulePromise(scheduler, req) {
  return new Promise ((resolve, reject) => {
    scheduler.updateJobSchedule(req, function(err, result) {
      if (err) { return reject(err);}
      resolve(result);
    });
  });
}
