// @sap/async-xsjs/lib/xsjs/jobs/Schedules.js
'use strict';

var assert = require('assert');
var Schedule = require('./Schedule');

module.exports = Schedules;

// internal properties are non-enumerable so they are not mixed with schedule ids

function Schedules(scheduler, jobId) {
  assert(scheduler, 'Missing scheduler');
  assert(jobId, 'Missing jobId');

  Object.defineProperty(this, '_scheduler', {
    enumerable: false,
    writable: false,
    value: scheduler
  });
  Object.defineProperty(this, '_jobId', {
    enumerable: false,
    writable: false,
    value: jobId
  });
}

Object.defineProperty(Schedules.prototype, 'delete', {
  enumerable: false,
  value: deleteSchedule
});
Object.defineProperty(Schedules.prototype, 'add', {
  enumerable: false,
  value: addSchedule
});

async function deleteSchedule(params) {
  if (params.id) {
    var scheduleId = params.id;
    if (!this.hasOwnProperty(scheduleId)) {
      throw new Error('Schedule ID not found');
    }
    await deleteJobSchedulePromise(this._scheduler, {
      jobId: this._jobId,
      scheduleId: scheduleId
    });
    delete this[scheduleId];
    return true;
  } else {
    throw new Error('"id" is mandatory');
  }
}

async function addSchedule(parameters) {
  var description = parameters.description;
  var xscron = parameters.xscron;
  var parameter = parameters.parameter;

  var body = {};
  if (parameter) {
    if (typeof parameter === 'string') {
      parameter = JSON.parse(parameter);
    } else if (typeof parameter !== 'object') {
      throw new Error('Wrong input type for "parameter"');
    }
    body.data = parameter;
  }
  if (!xscron) {
    throw new Error('"xscron" is mandatory');
  } else if (typeof xscron !== 'string') {
    throw new Error('Wrong input type for "xscron"');
  } else {
    body.cron = xscron;
  }
  if (description) {
    if (typeof description !== 'string') {
      throw new Error('Wrong input type for "description"');
    } else {
      body.description = description;
    }
  }

  var reply = await createJobSchedulePromise(this._scheduler, {
    jobId: this._jobId,
    schedule: body
  });

  var nscheduleId = reply.scheduleId ;
  this[nscheduleId] = await new Schedule(this._scheduler, reply.description, reply.data,
    reply.cron, reply.active, this._jobId, nscheduleId);
  return nscheduleId;
}

function deleteJobSchedulePromise(scheduler, req) {
  return new Promise ((resolve, reject) => {
    scheduler.deleteJobSchedule(req, function(err){
      if (err) { return reject(err);}
      resolve();
    });
  });
}

function createJobSchedulePromise(scheduler, req) {
  return new Promise((resolve, reject) => {
    scheduler.createJobSchedule(req, function(err, result){
      if (err) { return reject(err);}
      resolve(result);
    });
  });
}
