// @sap/async-xsjs/lib/xsjs/db/common/DbOptions.js
'use strict';

var _ = require('lodash');
var VError = require('verror');
var HttpError = require('../../../utils/errors/HttpError');
var connOptions = require('@sap/hdbext').connectionOptions;
var instanceManager = require('@sap/instance-manager');

var INSTANCE_MANAGER_OPTS = [
  'user', 'password',
  'post_managed_instance_url', 'delete_managed_instance_url',
  'get_managed_instance_url', 'get_all_managed_instances_url',
  'polling_interval_millis', 'polling_timeout_seconds',
  'cache_max_items', 'cache_item_expire_seconds'
];

var SERVICE_MANAGER_OPTS = [
  'clientid', 'clientsecret',
  'sm_url', 'url','polling_interval_millis', 'polling_timeout_seconds',
  'cache_max_items', 'cache_item_expire_seconds'
];

var DEFAULT_OPTIONS = {
  pooling: true,
  maxPoolSize: 10,
  connectionLifetime: 30
};

module.exports = DbOptions;

function DbOptions(globalOptions, xsApplicationUser) {
  this._globalOptions = _.extend({}, DEFAULT_OPTIONS, globalOptions);
  this._xsApplicationUser = xsApplicationUser;
  this._instManager = null;
}

DbOptions.prototype._getInstanceManager = function (cb) {
  if (this._instManager) {
    return cb(null, this._instManager);
  }
  var self = this;
  instanceManager.create(this._globalOptions, function (err, instManager) {
    if (err) {
      return cb(err);
    }
    self._instManager = instManager;
    cb(null, instManager);
  });
};

DbOptions.prototype.forRequest = function (req, locale) {
  return new DbRequestOptions(this, req, locale);
};

function DbRequestOptions(dbOptions, req, locale) {
  this._dbOptions = dbOptions;
  this._globalOptions = _.cloneDeep(dbOptions._globalOptions);
  this._xsApplicationUser = dbOptions._xsApplicationUser;
  this._req = req;
  this._locale = locale;
}

DbRequestOptions.prototype.getSqlccOptions = function (sqlccName) {
  let options;
  if (isInstanceManagerOptions(this._globalOptions)) {
    options = _.omit(this._globalOptions, INSTANCE_MANAGER_OPTS);
  } else if (isServiceManagerOptions(this._globalOptions)) {
    options = _.omit(this._globalOptions, SERVICE_MANAGER_OPTS);
  } else {
    options = this._globalOptions;
  }
  if (!options.sqlcc || !options.sqlcc[sqlccName]) {
    throw new TypeError('SQLCC connection with name "' + sqlccName +
      '" not found in the provided hana configurations');
  }

  options = _.merge({}, options, options.sqlcc[sqlccName]);
  return addRequestOptions(this._req, this._locale, options, this._xsApplicationUser);
};

DbRequestOptions.prototype.getInstanceOptions = function (callback) {
  if (!isInstanceManagerOptions(this._globalOptions) && !isServiceManagerOptions(this._globalOptions)) {
    var options = addRequestOptions(this._req, this._locale, this._globalOptions, this._xsApplicationUser);
    return callback(null, options);
  }
  var self = this;
  this._dbOptions._getInstanceManager(function (err, instManager) {
    if (err) {
      return callback(err);
    }
    var tenant = getTenant(self._req.authInfo);
    if (!tenant) {
      return callback(new Error('Could not get HANA service instance as no tenant is available in current context.'));
    }
    instManager.get(tenant, function (err, instance) {
      if (err) {
        return callback(err);
      }
      if (!instance) {
        return callback(new VError("No service instance for tenant '%s'", tenant));
      }
      if (instance.status !== 'CREATION_SUCCEEDED') {
        return callback(new VError("Status of service instance for tenant '%s' is %s", tenant, instance.status));
      }
      var hanaOptions;
      if (isInstanceManagerOptions(self._globalOptions)){
        hanaOptions = _.merge(_.omit(self._globalOptions, INSTANCE_MANAGER_OPTS), instance.credentials);
      } else {
        hanaOptions = _.merge(_.omit(self._globalOptions, SERVICE_MANAGER_OPTS), instance.credentials);
      }
      callback(null, addRequestOptions(self._req, self._locale, hanaOptions, self._xsApplicationUser));
    });
  });
};

DbRequestOptions.prototype.getInstanceOptionsPromise = function () {
  var self = this;
  return new Promise ((resolve, reject) => {
    self.getInstanceOptions((err, result) => {
      if (err) { return reject(err); }
      resolve(result);
    });
  });
};

function getTenant(authInfo) {
  return authInfo && authInfo.getZoneId();
}

function getUserToken(authInfo) {
  return authInfo && authInfo.getHdbToken();
}

function isInstanceManagerOptions(options) {
  return !!options.get_managed_instance_url;
}

function isServiceManagerOptions(options) {
  return !!options.sm_url;
}

function addRequestOptions(req, locale, hanaSettings, xsApplicationUser) {
  var hana = _.merge({}, hanaSettings, connOptions.getRequestOptions(req));

  if (locale && locale.dbLocale) {
    hana.locale = locale.dbLocale;
  }

  if (xsApplicationUser === false) {
    delete hana['sessionVariable:XS_APPLICATIONUSER'];
  }

  if (hana.connectWithLoggedUser) {
    var userToken = getUserToken(req.authInfo);
    if (!userToken) {
      throw new HttpError(400, 'No user token in request');
    }
    hana.user = '';
    hana.password = userToken;
  }
  delete hana.sqlcc;
  return hana;
}
