// @sap/async-xsjs/lib/xsjs/db/dbapi/ResultSet.js
'use strict';

var argumentsUtil = require('../common/arguments-validation');
var ResultSetMetaData = require('./ResultSetMetaData');
var conversions = require('./conversions');

module.exports = ResultSet;

function ResultSet(setOfRows) {
  this._rows = setOfRows.rows;
  this._columnInfo = setOfRows.columnInfo;
  this._rowCursor = 0;
  this._row = null;
}
/**
 * Fetches the next row
 * @returns {boolean} True if successful
 */
ResultSet.prototype.next = function () {
  if (Array.isArray(this._rows) && this._rowCursor < this._rows.length) {
    this._row = this._rows[this._rowCursor++];
    return true;
  }
  this._rows = [];
  this._row = undefined;
  return false;
};

/**
 * Closes the ResultSet
 * @throws Throws an error if the object the method is being called on is not valid.
 */
ResultSet.prototype.close = function () {

};

/**
 * Checks if the ResultSet is closed.
 * @returns {boolean} Returns true if the ResultSet is already closed, false if not
 * @throws Throws an error if the object the method is being called on is not valid.
 */
ResultSet.prototype.isClosed = function () {
  return false;
};

/**
 * Returns the metadata of the result set
 * @returns {db.ResultSetMetaData} The metadata of the result set
 */
ResultSet.prototype.getMetaData = function () {
  return new ResultSetMetaData(this._columnInfo);
};

/**
 * Returns an integer value of the specified column, for TINYINT, SMALLINT, INT, and BIGINT column types.
 * An exception is thrown if the value is bigger than 9007199254740992 (2^53) or smaller than -9007199254740992 (-2^53).
 * @param {integer} columnIndex The target column <b>starting from 1</b>
 * @returns {integer} Integer
 * @throws Throws an error if the index parameter is not valid.
 */
ResultSet.prototype.getInteger = function (columnIndex) {
  argumentsUtil.validateGetterArgs(arguments, 1, this._row);
  return conversions.getInteger(this._row[columnIndex - 1]);
};

ResultSet.prototype.getInt = ResultSet.prototype.getInteger;

/**
 * Returns a ctypes.Int64 value of the specified column. getBigInt is used for BIGINT column types.
 * @param {integer} columnIndex The target column, <b>starting from 1</b>
 * @returns {ctypes.Int64}
 * @throws Throws an error if the index parameter is not valid.
 */
ResultSet.prototype.getBigInt = function (columnIndex) {
  argumentsUtil.validateGetterArgs(arguments, 1, this._row);
  return conversions.getBigInt(this._row[columnIndex - 1]);
};

/**
 * Returns a number value of the specified column. getDecimal is used for DECIMAL column types.
 * @param {integer} columnIndex The target column <b>starting from 1</b>
 * @returns {number} Number representation
 * @throws Throws an error if the index parameter is not valid.
 */
ResultSet.prototype.getDecimal = function (columnIndex) {
  argumentsUtil.validateGetterArgs(arguments, 1, this._row);
  return conversions.getDecimal(this._row[columnIndex - 1]);
};

/**
 * Returns a number value of the specified column. getDouble is used for DOUBLE column types.
 * @param {integer} columnIndex The target column <b>starting from 1</b>
 * @returns {number} Number representation
 * @throws Throws an error if the index parameter is not valid.
 */
ResultSet.prototype.getDouble = ResultSet.prototype.getDecimal;

/**
 * Returns a number value of the specified column. getReal is used for REAL column types.
 * @param {integer} columnIndex The target column <b>starting from 1</b>
 * @returns {number} Number representation
 * @throws Throws an error if the index parameter is not valid.
 */
ResultSet.prototype.getReal = ResultSet.prototype.getDecimal;

/**
 * Returns a number value of the specified column. getFloat is used for FLOAT column types.
 * @param {integer} columnIndex The target column <b>starting from 1</b>
 * @returns {number} Number representation
 * @throws Throws an error if the index parameter is not valid.
 */
ResultSet.prototype.getFloat = ResultSet.prototype.getDecimal;

/**
 * Returns a ArrayBuffer value of the specified column. getBlob is used for BLOB column types.
 * @param {integer} columnIndex The target column, <b>starting from 1</b>
 * @returns {ArrayBuffer} Blob representation
 * @throws Throws an error if the index parameter is not valid.
 */
ResultSet.prototype.getBlob = function (columnIndex) {
  argumentsUtil.validateGetterArgs(arguments, 1, this._row);
  return conversions.getBlob(this._row[columnIndex - 1],
    this._columnInfo[columnIndex - 1].nativeType);
};

/**
 * Returns a string value of the specified column. getClob is used for CLOB column types.
 * @param {integer} columnIndex The target column <b>starting from 1</b>
 * @returns {string} String representation
 * @throws Throws an error if the index parameter is not valid.
 */
ResultSet.prototype.getClob = function (columnIndex) {
  argumentsUtil.validateGetterArgs(arguments, 1, this._row);
  return conversions.getClob(this._row[columnIndex - 1],
    this._columnInfo[columnIndex - 1].nativeType);
};

/**
 * Returns a string value of the specified column. getNClob is used for NCLOB and TEXT column types.
 * @param {integer} columnIndex The target column <b>starting from 1</b>
 * @returns {string} String representation
 * @throws Throws an error if the index parameter is not valid.
 */
ResultSet.prototype.getNClob = ResultSet.prototype.getClob;

/**
 * Returns a string value of the specified column. getString is used for CHAR and VARCHAR column types. ASCII only, not suitable for strings containing unicode characters.
 * @param {integer} columnIndex The target column <b>starting from 1</b>
 * @returns {string} String representation
 * @throws Throws an error if the index parameter is not valid.
 */
ResultSet.prototype.getString = function (columnIndex) {
  argumentsUtil.validateGetterArgs(arguments, 1, this._row);
  return conversions.getString(this._row[columnIndex - 1],
    this._columnInfo[columnIndex - 1].nativeType);
};

/**
 * Returns a string value of the specified column. getText is used for TEXT column types.
 * @param {integer} columnIndex The target column <b>starting from 1</b>
 * @returns {string} String representation
 * @throws Throws an error if the index parameter is not valid.
 */
ResultSet.prototype.getText = ResultSet.prototype.getClob;

/**
 * Returns a string value of the specified column. getNString is used for NCHAR, NVARCHAR, SHORTTEXT column types.
 * @param {integer} columnIndex The target column <b>starting from 1</b>
 * @returns {string} String representation
 * @throws Throws an error if the index parameter is not valid.
 */
ResultSet.prototype.getNString = ResultSet.prototype.getString;

/**
 * Returns an ArrayBuffer object of the specified column. getBString is used for BINARY and VARBINARY column types.
 * @param {integer} columnIndex The target column <b>starting from 1</b>
 * @returns {ArrayBuffer} value ArrayBuffer object
 * @throws Throws an error if the index parameter is not valid or the SQL type of the queried parameter does not match.
 */
ResultSet.prototype.getBString = ResultSet.prototype.getBlob;

/**
 * Used to retrieve the value of a DATE column in a ResultSet.
 * @param {integer} columnIndex The index of the column in the resultset <b>starting from 1</b>
 * @returns {Date} A JavaScript Date object representing the value
 * @throws Throws an error if the index parameter is not valid.
 */
ResultSet.prototype.getDate = function (columnIndex) {
  argumentsUtil.validateGetterArgs(arguments, 1, this._row);
  return conversions.getDate(this._row[columnIndex - 1]);
};

/**
 * Used to retrieve the value of a TIME column in a ResultSet.
 * @param {integer} columnIndex The index of the column in the resultset <b>starting from 1</b>
 * @returns {Date} A JavaScript Date object representing the value
 * @throws Throws an error if the index parameter is not valid.
 */
ResultSet.prototype.getTime = function (columnIndex) {
  argumentsUtil.validateGetterArgs(arguments, 1, this._row);
  return conversions.getTime(this._row[columnIndex - 1]);
};

/**
 * Used to retrieve the value of a SECONDDATE column in a ResultSet.
 * @param {integer} columnIndex The index of the column in the resultset <b>starting from 1</b>
 * @returns {Date} A JavaScript Date object representing the value
 * @throws Throws an error if the index parameter is not valid.
 */
ResultSet.prototype.getSeconddate = ResultSet.prototype.getDate;

/**
 * Used to retrieve the value of a TIMESTAMP column in a ResultSet. <br>
 * As this type contains only time information and no date, the JavaScript's date object will always be 1 Jan 1970 plus the time offset.<br>
 * For example: if the stored value is 10:00:00, the JavaScript Date object will be: 1 Jan 1970 10:00:00.
 * @param {integer} columnIndex The index of the column in the resultset <b>starting from 1</b>
 * @returns  {Date} A JavaScript Date object representing the value<br>
 * @throws Throws an error if the index parameter is not valid.
 */
ResultSet.prototype.getTimestamp = ResultSet.prototype.getDate;
